### PR TITLE
feat: SDLF DDK MWAA pipeline orchestration

### PR DESCRIPTION
**Pull request by devansh-gandhi**
### Description

With the addition of the MWAAEnvironment and MWAATrigerDagsStage to aws-ddk, an example is created to show case SDLF DDK LightWeight with MWAA orchestration. 

### Changes Made

• A new parameter called orchestration is added to the parameters.json to determine the orchestration type (SFN/ MWAA). If the parameter is set to SFN, the existing pipeline stages will be deployed. However, when it is set to MWAA, a MWAATriggerDagsStage and MWAAEnvironment is deployed instead of the stage-a and stage-b step functions. 
• New Stage-a airflow dag is created to trigger Stage-a lambda functions. Stage-b changes in progress. 

• Error handling logic is added to the dag to handle processing failures.

The following changes are implemented in the current pipeline structure -

• In routing-a lambda function, the event will now be passed to the step function as a dict instead of string
• Within the SDLFLightTransform stack, the creation the lambda functions is decoupled from creation of the step function tasks. This seperation enables the lambda functions to be created independently for the MWAA orchestration and prevents unnecessary creation of the step function task.
• Conditional statements added to ensure only resources required for that orchaestration type are deployed

Note: Same parameter values are used for the step functions(stage-a/stage-b) and MWAA. So depending on the orchestration, it will run either sfn or mwaa pipeline but cannot run both.

### Related Issues

#54

### CODE REVIEW

**Comment by devansh-gandhi** [comment]
Hi @anmolsgandhi , I have created stage-a of SDLF DDK Lightweight with MWAA pipeline orchestration. Could you please review before I create stage-b ? 

Thanks!

**Comment by anmolsgandhi** [comment]
Hi Devansh, Thanks for opening the PR. From the changes made section above, i had a similar idea on implementing it. I think it might be a good idea to get stageB in there as i understand it will only be a dag addition. Can we also get an architecture diagram to showcase changes and also update readme details wherever required. This will help to review the usecase patten end to end holistically. Looking forward to looking into this!

**Comment by anmolsgandhi** [comment]
Hi @devansh-gandhi, checking in to see if you need any help. Let us know if you need to discuss design or implementation.

**Comment by devansh-gandhi** [comment]
Hi @anmolsgandhi , @malachi-constant, I am ready with the stage-B implementation. Could you please review the solution? 

I have made the following changes in latest commits-

1) Updated sdlf_heavy_transform.py, base_stack.py and routing_lambda.py for stage-B transformations
2) Added dag for stage-B transformations
3) Updated base_stack.py to enable separate MWAAEnvironment for each team
4) restructured dags folder and updated IAM statements in create_mwaa_env function to support multiple team
5) The creation of MWAAEnvironment moved to standard_pipeline stack to ensure MWAAEnvironment gets created only for standard pipelines

File: sdlf-ddk-lightweight/python/data_lake/pipelines/common_stages/sdlf_heavy_transform.py
Comment (malachi-constant) [inline sdlf-ddk-lightweight/python/data_lake/pipelines/common_stages/sdlf_heavy_transform.py:38]: Do we want to default this to `sfn`?

File: sdlf-ddk-lightweight/python/data_lake/pipelines/common_stages/sdlf_light_transform.py
Comment (malachi-constant) [inline sdlf-ddk-lightweight/python/data_lake/pipelines/common_stages/sdlf_light_transform.py:40]: Same as above

File: sdlf-ddk-lightweight/python/data_lake/pipelines/sdlf_base_stack.py
Comment (malachi-constant) [inline sdlf-ddk-lightweight/python/data_lake/pipelines/sdlf_base_stack.py:91]: Maybe leave this as default behavior if not in config

**Review by malachi-constant — COMMENTED** [review]
Thanks for the contribution! See my notes on default `orchestration` and would like to get this fully tested before approval.

**Comment by anmolsgandhi** [comment]
@devansh-gandhi Adding on to lucas's point, can you comment on the various usecases you have tested this solution? we can then evaluate if there are any missing end cases and then do end to end testing for all cases.

File: sdlf-ddk-lightweight/python/data_lake/pipelines/sdlf_base_stack.py
Comment (devansh-gandhi) [reply sdlf-ddk-lightweight/python/data_lake/pipelines/sdlf_base_stack.py:91]: Hi, I can make that change. Could you please confirm if i should set the orchestration default in the SDLFHeavyTransformConfig / SDLFLightTransformConfig or leave it in the base_stack ?

**Comment by devansh-gandhi** [comment]
Hi @anmolsgandhi, I have tested the following use cases using the files from the examples folder - 

1) 1 team,1 dataset -> orchestration set to either SFN/MWAA ->  Sucessful execution
2) 1 team, 2 datasets -> orchestration set to either SFN/MWAA ->  Sucessful execution  
3) 1 team, 2 datasets -> orchestration set to SFN and MWAA -> Failed execution for SFN, MWAA is triggered even for SFN. 1 Team cannot have both SFN and MWAA orchestration
4) 2 teams, 2 datasets(one each) -> Same orchestration either SFN/MWAA ->  Sucessful execution   
5) 2 teams, 2 datasets(one each) -> Different orchestration between teams->  Sucessful execution   
6) 2 teams, 3 datasets  -> same and different orchestration between teams  ->  Sucessful execution

**Comment by malachi-constant** [comment]
@devansh-gandhi Did you test with `cicd_enabled=True`. As that's set in the configuration I believe we need to update `app.py` to pass `ec2:DescribeAvailbilityZones` to the synth stage.

```
...
pipeline.add_synth_action(
        role_policy_statements=[
            iam.PolicyStatement.from_json(
                {"Action": "ec2:DescribeAvailabilityZones", "Resource": "*"}
            )
        ]
    )
...
```

File: sdlf-ddk-lightweight/python/requirements.txt
Comment (malachi-constant) [inline sdlf-ddk-lightweight/python/requirements.txt:1]: Can we remove the version pin to use latest, just tested with `1.3.1`

**Review by malachi-constant — CHANGES_REQUESTED** [review]
Testing seemed to work fine, see my comments on DDK version and CI/CD support for the MWAA orchestration mode of this SDLF example.

File: sdlf-ddk-lightweight/python/requirements.txt
Comment (devansh-gandhi) [reply sdlf-ddk-lightweight/python/requirements.txt]: Hi, 

I have updated the pipeline synth action in app.py to pass ec2:DescribeAvailbilityZones to the synth stage. I had tested it by passing it as a flag in cdk.json context but did not push that as I was not sure. 

Regarding the aws_ddk_core version , I have changed that to 1.3.1. Please let me if I should remove it all together.

**Review by malachi-constant — APPROVED** [review]

**Pull request merged**
